---
title: "Dengue RPL project"
output: html_notebook
---

### Intialization and installation
```{r, echo=F, eval=F, message=F}
### Initialize packages
library(xtail)
library(riborex)
library(rhdf5)
library(dplyr)
library(plyr)
library(GenomicRanges)
library(rtracklayer)
library(ggplot2)
library(parallel)
library(Biostrings)
library(Rsamtools)
library(tidyverse)
library(feather)
library(DESeq2)
library(DEGseq)
library(data.table)
library(GenomicAlignments)
library(GGally)
```

### Read in annotations and GFF data
```{r}
# Read in Appris list
appris <- read_tsv("/data/riboseq/Supp/seqs/human/human_appris_data.principal.txt",col_names = F)

# Get a list of all principal transcripts
princ_app <- appris %>% filter(substr(X5,1,1)=="P")
princ_app <- princ_app[order(princ_app$X1),]

# Pick the first principal transcript in genes with multiple principal transcripts 
all_enstt <- princ_app[!duplicated(princ_app$X1),]

# Create a list of mapped names
gene_trans_map <- c(all_enstt$X1,"DENGUE")
names(gene_trans_map) <- c(all_enstt$X3,"DENGUE")

# Read in GFF file
gff <- readGFFAsGRanges("/data/riboseq/Supp/seqs/Virus/Dengue/human_dengue.gff3")
tlist <- unique(gff$Name)
glist <- gene_trans_map[tlist]

gff_alllen <- width(gff)
names(gff_alllen) <- seqnames(gff)
gff_cdslen <- width(gff[gff$type=="CDS"])
names(gff_cdslen) <- seqnames(gff[gff$type=="CDS"])

glen <- sapply(tlist,function(x){sum(gff_alllen[names(gff_alllen)==x])})
```


### Read in count data from bamFile (fast)
```{r}
datasets <- c("RiboNSC1","RiboNSC2","RiboNSC3","RibosiP1","RibosiP2","RibosiP3",
              "RNANSC1","RNANSC2","RNANSC3","RNAsiP1","RNAsiP2","RNAsiP3")
filepath <- "/data/riboseq/Dengue/output/"

bcountData <- data.frame(tid=tlist, gene=glist, matrix(NA,ncol=length(datasets),nrow = length(tlist)))
colnames(bcountData)[3:(length(datasets)+2)] <- datasets

for(fn in datasets){
  bamData <- readGAlignments(paste0(filepath,fn,".bam"))
  bamData <- bamData[strand(bamData)=="+" & qwidth(bamData)>=10 & qwidth(bamData)<=50]
  bamCounts <- table(seqnames(bamData))
  bcountData[fn] <- as.integer(bamCounts[as.character(bcountData$tid)])
}

bcountData$Length <- glen[as.character(bcountData$tid)]
bcountData$CDSLength <- gff_cdslen[as.character(bcountData$tid)]

write.table(bcountData,paste0(filepath,"shah/all_counts_bam.tsv"),quote=F,sep="\t",row=F,col=T)

# Calcualte TPM
btpmData <- bcountData

for(i in 9:14){
  btpmData[,i] <- bcountData[,i]*1e3/bcountData$Length
  btpmData[,i] <- btpmData[,i]*1e6/sum(btpmData[,i])
}
for(i in 3:8){
  btpmData[,i] <- bcountData[,i]*1e3/bcountData$CDSLength
  btpmData[,i] <- btpmData[,i]*1e6/sum(btpmData[,i])
}

# Calcualte RPKM
brpkmData <- bcountData

for(i in 9:14){
  brpkmData[,i] <- bcountData[,i]*1e6/sum(bcountData[,i])
  brpkmData[,i] <- brpkmData[,i]*1e3/bcountData$Length
}
for(i in 3:8){
  brpkmData[,i] <- bcountData[,i]*1e6/sum(bcountData[,i])
  brpkmData[,i] <- brpkmData[,i]*1e3/bcountData$CDSLength
}

write.table(btpmData,paste0(filepath,"shah/all_tpms_bam.tsv"),quote=F,sep="\t",row=F,col=T)
write.table(brpkmData,paste0(filepath,"shah/all_rpkms_bam.tsv"),quote=F,sep="\t",row=F,col=T)
rm(bamData)
```

### Distribution of read-lengths
```{r}
len_dist <- data.frame(data = rep(datasets,each=length(10:50)),
                       length = rep(10:50,length(datasets)),
                       counts = NA)
for(fn in datasets){
  bamData <- readGAlignments(paste0(filepath,fn,".bam"))
  bamData <- bamData[strand(bamData)=="+" & qwidth(bamData)>=10 & qwidth(bamData)<=50]
  len_dist[len_dist$data==fn,"counts"] <- as.integer(table(factor(qwidth(bamData),levels=10:50)))
}
len_dist_plot <- ggplot(len_dist, aes(x=length, y=counts)) + geom_line() + facet_wrap(~data, scales="free")
ggsave(len_dist_plot, filename = paste0(filepath,"length_dist.png"), width=10, height=10)

write_tsv(len_dist, path=paste0(filepath,"length_dist.tsv"))
```

### 3nt periodicity
```{r}
genes <- as.character(bcountData$tid)
dataset <- "data"
out_data <- c()

get_nt_period <- function(fid,gene,dataset,left,right){
  data_mat <- H5Dread(H5Dopen(fid,paste0("/",gene,"/",dataset,"/reads/data")))  # Get the matrix of read counts
  pos_sum <- colSums(data_mat)  # Position-specific sum of reads of all lengths
  pos_sum <- pos_sum[left:(ncol(data_mat)-right)] # Ignore reads in parts of the buffer region defined by left/right
  return(pos_sum)
}

n_buffer <- 25  # Number of nts in buffer to be included in the summary table/plot
n_gene <- 50  # Number of nts in CDS to be included in the summary table/plot
n_total <- n_buffer+n_gene

# NOTE: Do not use mclapply when accessing H5 data

# Get gene and position specific total counts of all lengths
for(fn in datasets){
  fid <- H5Fopen(paste0(filepath,fn,".h5"))

  gene_sp_pos_counts <- lapply(genes,function(gene){
                              utr5 <- width(gff[gff$type=="UTR5" & gff$Name==gene])+1
                              utr3 <- width(gff[gff$type=="UTR3" & gff$Name==gene])
                              get_nt_period(fid=fid,
                                            gene=gene,
                                            dataset=dataset,
                                            left=utr5,
                                            right=utr3)
                        })
  x <- lapply(gene_sp_pos_counts,head,n=n_total) # Subset reads mapping to a fixed region across genes (5')
  y <- lapply(gene_sp_pos_counts,tail,n=n_total) # Subset reads mapping to a fixed region across genes (3')
  a <- sapply(x,length)
  x <- x[a==n_total]
  y <- y[a==n_total]

  lsum <- colSums(matrix(unlist(x),byrow=T,nrow=length(x))) # Position-specific sums of reads (5')
  rsum <- colSums(matrix(unlist(y),byrow=T,nrow=length(x))) # Position-specific sums of reads (3')

  # Create a dataframe to store the output for plots/analyses
  out_data <- rbind(out_data, data.frame(Data = fn,
                                        Pos = rep(seq(-n_buffer+1,n_gene,1),2),
                                        Counts = c(lsum,rsum),
                                        End = rep(c("5'","3'"),each=n_total)))
  H5Fclose(fid)
}

period_plot <- ggplot(out_data, aes(x=Pos, y=Counts, col=Data)) + geom_line() + facet_wrap(~End, scales="free") + scale_y_log10()
ggsave(period_plot, filename = paste0(filepath,"periodicity_plot.png"), width=8, height=5)

write_tsv(out_data, path=paste0(filepath,"periodicity_3n.tsv"))

```

### Dengue plot
```{r}
gene <- "DENGUE"
dtype <- rep(c("RPF","mRNA"),each=6)
etype <- rep(rep(c("NSC","siP"),each=3),2)
names(dtype) <- datasets
names(etype) <- datasets
libsize <- colSums(bcountData[,3:14])

dengue_pos_read <- c()
for(fn in datasets){
  fid <- H5Fopen(paste0(filepath,fn,".h5"))
  d <- colSums(H5Dread(H5Dopen(fid,paste0("/",gene,"/",dataset,"/reads/data"))))
  dengue_pos_read <- rbind(dengue_pos_read, data.frame(data=fn,
                                                       pos=1:length(d),
                                                       counts=d,
                                                       ncounts=d/libsize[fn],
                                                       dtype=dtype[fn],
                                                       etype=etype[fn]))
  H5Fclose(fid)
}

# ggplot(dengue_pos_read, aes(x=pos, y=counts, col=data)) + geom_line() + scale_y_log10() + facet_wrap(dtype~etype)
rpf_pos_reads <- ggplot(dengue_pos_read %>% filter(dtype=="RPF"), aes(x=pos, y=counts, col=data)) + geom_line() + facet_wrap(~data,nrow = 6, scales="free")
rna_pos_reads <- ggplot(dengue_pos_read %>% filter(dtype=="mRNA"), aes(x=pos, y=counts, col=data)) + geom_line() + facet_wrap(~data,nrow = 6, scales="free")

ggsave(rpf_pos_reads, filename = paste0(filepath,"rpf_pos_reads.png"), width=10, height=10)
ggsave(rna_pos_reads, filename = paste0(filepath,"rna_pos_reads.png"), width=10, height=10)

```

```{r}
a <- tbl_df(dengue_pos_read) %>% group_by(data) %>% mutate(rollm=roll_mean(ncounts,50,fill=0))

sub_dengue <- a %>% 
  ungroup(data) %>% 
  group_by(pos,etype,dtype) %>%
  summarise(mean=mean(rollm))

ggplot(sub_dengue, aes(x=pos,y=mean,col=etype)) + geom_line() + facet_wrap(~dtype,nrow = 2) + scale_y_log10()
ggplot(a %>% filter(pos<=700), aes(x=pos,y=rollm,col=data)) + geom_line() + facet_wrap(~dtype,nrow = 2) + scale_y_log10()
ggplot(sub_dengue %>% filter(pos<=700), aes(x=pos,y=mean,col=etype)) + geom_line() + facet_wrap(~dtype,nrow = 2) + scale_y_log10()

ggplot(a %>% filter(pos>=9000), aes(x=pos,y=rollm,col=data)) + geom_line() + facet_wrap(~dtype,nrow = 2) + scale_y_log10()
ggplot(sub_dengue %>% filter(pos>=9000), aes(x=pos,y=mean,col=etype)) + geom_line() + facet_wrap(~dtype,nrow = 2) + scale_y_log10()

```


<!-- ### Plots -->
<!-- ### Subset of lm_eqn_all function to fit equations on ggplot -->
<!-- ```{r} -->
<!-- lm_eqn = function(df){ -->
<!--     df <- df[df$rna_counts!=0 & df$rpf_counts!=0 & !is.na(df$rna_counts) & !is.na(df$rpf_counts),] -->
<!--     m = lm(log10(df$rpf_counts) ~ log10(df$rna_counts)) -->
<!--     eq <- substitute(italic(R)^2~"="~r2,  -->
<!--          list(r2 = format(summary(m)$r.squared, digits = 2))) -->
<!--     as.character(as.expression(eq));                  -->
<!-- } -->
<!-- ``` -->

<!-- # Relationship between transcriptional and translational regulation -->
<!-- ```{r} -->
<!-- bcount_mod <- bcountData %>%  -->
<!--   gather(RPF,rpf_counts,c(3:6))  -->
<!-- bcount_mod <- bcount_mod %>% -->
<!--   gather(RNA,rna_counts,c(3:6)) -->

<!-- # cors <- ddply(bcount_mod,.(RNA,RPF),summarize,cor=format(signif(cor(rna_counts,rpf_counts),2),scientific=-2)) -->
<!-- cors2 <- ddply(bcount_mod,.(RNA,RPF),lm_eqn) -->

<!-- rpf_rna_plot <- ggplot(bcount_mod, aes(rna_counts, rpf_counts)) +  -->
<!--   geom_bin2d(binwidth=0.07) +  -->
<!--   scale_alpha(guide = "none") +  -->
<!--   scale_x_log10() +  -->
<!--   scale_y_log10() +  -->
<!--   # geom_smooth(method="lm") + -->
<!--   facet_grid(RPF ~ RNA) + -->
<!--   geom_text(data=cors2,aes(x = 5, y = 1e5,label=V1), parse = TRUE, inherit.aes=FALSE) + -->
<!--   labs(x="RNA counts", y= "RPF counts")  -->

<!-- rpf_rna_plot -->
<!-- ggsave(rpf_rna_plot, filename = paste0(filepath,"plots/rpf_rna_plot.png"), width = 10,height = 9) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- source("~/R/my_scripts/ggplot2/generic_ggplot_funcs.R") -->
<!-- rpf_cor_plot <- ggpairs(bcountData[,c(3:6)], aes(alpha=0.2), upper = list(continuous = my_custom_cor_color)) -->
<!-- rpf_cor_plot_log <- ggpairs_logarithmize(rpf_cor_plot) -->
<!-- rpf_cor_plot_log -->

<!-- ggsave(rpf_cor_plot_log, filename = paste0(filepath,"plots/rpf_cor_plot.png"), width = 10,height = 10) -->

<!-- rna_cor_plot <- ggpairs(bcountData[,c(7:10)], aes(alpha=0.2), upper = list(continuous = my_custom_cor_color)) -->
<!-- rna_cor_plot_log <- ggpairs_logarithmize(rna_cor_plot) -->
<!-- rna_cor_plot_log -->

<!-- ggsave(rna_cor_plot_log, filename = paste0(filepath,"plots/rna_cor_plot.png"), width = 10,height = 10) -->
<!-- ``` -->

<!-- ### DEGseq analyses -->
<!-- ```{r} -->
<!-- a <- (rowMaxs(as.matrix(bcountData[,3:10]))>=64 & (rowMins(as.matrix(bcountData[,3:10]))>0) | rowMeans(as.matrix(bcountData[,3:10]))>64) -->
<!-- high_cov_data <- bcountData[a,3:10] -->
<!-- high_cov_data$name <- bcountData$gene[a] -->
<!-- # high_cov_tpm <- as.matrix(btpmData[a,3:10]) -->

<!-- DEGexp(geneExpMatrix1 = high_cov_data[,c("name","ctl_fp")], geneCol1 = 1, expCol1 = 2, groupLabel1 = "ctl_fp", -->
<!--        geneExpMatrix2 = high_cov_data[,c("name","lin_fp")], geneCol2 = 1, expCol2 = 2, groupLabel2 = "lin_fp",  -->
<!--        outputDir = "/data/riboseq/Rustgi/output/ctl_lin_fp", thresholdKind = 5, method="MARS", foldChange=2, qValue = 0.01) -->

<!-- DEGexp(geneExpMatrix1 = high_cov_data[,c("name","ctl_rn")], geneCol1 = 1, expCol1 = 2, groupLabel1 = "ctl_rn", -->
<!--        geneExpMatrix2 = high_cov_data[,c("name","lin_rn")], geneCol2 = 1, expCol2 = 2, groupLabel2 = "lin_rn",  -->
<!--        outputDir = "/data/riboseq/Rustgi/output/ctl_lin_rn", thresholdKind = 5, method="MARS", foldChange=2, qValue = 0.01) -->

<!-- DEGexp(geneExpMatrix1 = high_cov_data[,c("name","ctl_fp")], geneCol1 = 1, expCol1 = 2, groupLabel1 = "ctl_fp", -->
<!--        geneExpMatrix2 = high_cov_data[,c("name","ctl_imp_fp")], geneCol2 = 1, expCol2 = 2, groupLabel2 = "ctl_imp_fp",  -->
<!--        outputDir = "/data/riboseq/Rustgi/output/ctl_ctl_imp_fp", thresholdKind = 5, method="MARS", foldChange=2, qValue = 0.01) -->

<!-- DEGexp(geneExpMatrix1 = high_cov_data[,c("name","ctl_rn")], geneCol1 = 1, expCol1 = 2, groupLabel1 = "ctl_rn", -->
<!--        geneExpMatrix2 = high_cov_data[,c("name","ctl_imp_rn")], geneCol2 = 1, expCol2 = 2, groupLabel2 = "ctl_imp_rn",  -->
<!--        outputDir = "/data/riboseq/Rustgi/output/ctl_ctl_imp_rn", thresholdKind = 5, method="MARS", foldChange=2, qValue = 0.01) -->

<!-- DEGexp(geneExpMatrix1 = high_cov_data[,c("name","lin_fp")], geneCol1 = 1, expCol1 = 2, groupLabel1 = "lin_fp", -->
<!--        geneExpMatrix2 = high_cov_data[,c("name","lin_imp_fp")], geneCol2 = 1, expCol2 = 2, groupLabel2 = "lin_imp_fp",  -->
<!--        outputDir = "/data/riboseq/Rustgi/output/lin_lin_imp_fp", thresholdKind = 5, method="MARS", foldChange=2, qValue = 0.01) -->

<!-- DEGexp(geneExpMatrix1 = high_cov_data[,c("name","lin_rn")], geneCol1 = 1, expCol1 = 2, groupLabel1 = "lin_rn", -->
<!--        geneExpMatrix2 = high_cov_data[,c("name","lin_imp_rn")], geneCol2 = 1, expCol2 = 2, groupLabel2 = "lin_imp_rn",  -->
<!--        outputDir = "/data/riboseq/Rustgi/output/lin_lin_imp_rn", thresholdKind = 5, method="MARS", foldChange=2, qValue = 0.01) -->

<!-- ``` -->

<!-- ```{r} -->
<!-- # Read in differential mRNA/RPF data from DEGseq analyses. -->
<!-- comparisons <- c("ctl_lin","ctl_ctl_imp","lin_lin_imp") -->
<!-- all_fc_data <- c() -->

<!-- for(i in 1:3){ -->
<!--   str_m <- fread(paste0("/data/riboseq/Rustgi/output/",comparisons[i],"_rn/output_score.txt"),h=T) -->
<!--   str_r <- fread(paste0("/data/riboseq/Rustgi/output/",comparisons[i],"_fp/output_score.txt"),h=T) -->

<!--   str_m <- str_m %>% -->
<!--           select(GeneNames, `log2(Fold_change) normalized`, `q-value(Storey et al. 2003)`) -->
<!--   str_r <- str_r %>% -->
<!--           select(GeneNames, `log2(Fold_change) normalized`, `q-value(Storey et al. 2003)`) -->
<!--   names(str_m) <- c("Gene","FC_mRNA","SigM") -->
<!--   names(str_r) <- c("Gene","FC_RPF","SigR") -->
<!--   str_all <- full_join(str_m,str_r,by="Gene") -->
<!--   str_all$comp <- comparisons[i] -->
<!--   all_fc_data <- bind_rows(all_fc_data,str_all) -->
<!-- } -->
<!-- all_fc_data$FC_mRNA <- -all_fc_data$FC_mRNA -->
<!-- all_fc_data$FC_RPF <- -all_fc_data$FC_RPF -->
<!-- all_fc_data$SigM <- (all_fc_data$SigM<0.001) -->
<!-- all_fc_data$SigR <- (all_fc_data$SigR<0.001) -->

<!-- ``` -->

<!-- ### 3. Identify nature of transcriptional and translational reglulation. -->
<!-- ```{r} -->
<!-- # Identify change categories -->
<!-- all_fc_data$Change <- NA -->
<!-- all_fc_data$Change[abs(all_fc_data$FC_mRNA)>1 & abs(all_fc_data$FC_RPF-all_fc_data$FC_mRNA)<=1] <- "Transcription only" -->
<!-- all_fc_data$Change[abs(all_fc_data$FC_RPF)>1 & abs(all_fc_data$FC_mRNA)<=1] <- "Translation only" -->
<!-- all_fc_data$Change[abs(all_fc_data$FC_mRNA)>1 & abs(all_fc_data$FC_RPF-all_fc_data$FC_mRNA)>1] <- "Translation reinforcement" -->
<!-- all_fc_data$Change[abs(all_fc_data$FC_mRNA)>1 & abs(all_fc_data$FC_RPF)>1 & sign(all_fc_data$FC_mRNA)==sign(all_fc_data$FC_RPF) & abs(all_fc_data$FC_mRNA-all_fc_data$FC_RPF)>1 & abs(all_fc_data$FC_mRNA)>abs(all_fc_data$FC_RPF)] <- "Translation buffering (partial)" -->
<!-- all_fc_data$Change[abs(all_fc_data$FC_mRNA)>1 & abs(all_fc_data$FC_RPF)<=1] <- "Translation buffering (complete)" -->
<!-- all_fc_data$Change[abs(all_fc_data$FC_mRNA)>1 & abs(all_fc_data$FC_RPF)>1 & sign(all_fc_data$FC_mRNA)!=sign(all_fc_data$FC_RPF)] <- "Translation antagonism" -->
<!-- all_fc_data$Change[abs(all_fc_data$FC_mRNA)<=1 & abs(all_fc_data$FC_RPF)<=1] <- "No change" -->
<!-- all_fc_data$Change <- as.factor(all_fc_data$Change) -->
<!-- all_fc_data$Change <- factor(all_fc_data$Change,levels(all_fc_data$Change)[c(1,2,6,7,5,4,3)]) -->

<!-- all_fc_data <- all_fc_data[!is.na(all_fc_data$Change),] -->

<!-- write.table(all_fc_data,paste0(filepath,"output/all_mrna_rpf_foldchange.tsv"),quote=F,sep="\t",row=F,col=T) -->

<!-- comparisons <- c("ctl_lin","ctl_ctl_imp","lin_lin_imp") -->
<!-- # comparisons_text <- c(expression(paste("wt vs ",Delta,italic("lin28b"))),  -->
<!-- #                       expression(paste("wt vs ",Delta,italic("imp1"))), -->
<!-- #                       expression(paste(Delta,italic("lin28b")," vs ",Delta,italic("lin28b"),Delta,italic("imp1")))) -->
<!-- comparisons_text <- c("WT vs LIN",  -->
<!--                       "WT vs IMP", -->
<!--                       "LIN vs LIN-IMP") -->


<!-- for(i in 1:3){ -->
<!--   m <- lm(all_fc_data$FC_RPF[all_fc_data$comp==comparisons[i]] ~ all_fc_data$FC_mRNA[all_fc_data$comp==comparisons[i]]) -->
<!--   mrna_rpf_fc_change <- ggplot(filter(all_fc_data,comp==comparisons[i]), aes(x=FC_mRNA, y=FC_RPF, col=Change)) + -->
<!--     geom_point(alpha=0.5) + -->
<!--     scale_alpha(guide = 'none') +  -->
<!--     scale_color_brewer(palette = "Set1") + -->
<!--     theme(legend.title=element_blank()) +  -->
<!--     labs(x="mRNA log2 fold-change", y= "RPF log2 fold-change") +  -->
<!--     guides(colour = guide_legend(override.aes = list(size=5,aplha=0.8))) + -->
<!--     xlim(-9.5,9.5) + -->
<!--     ylim(-9.5,9.5) + -->
<!--     coord_fixed(ratio = 1) + -->
<!--     annotate("text", x = -5, y = 9, label = comparisons_text[i]) + -->
<!--     annotate("text", x = -5, y = 7.5, label = r2plot(m), parse=TRUE) -->

<!--   mrna_rpf_fc_change -->
<!--   ggsave(mrna_rpf_fc_change, filename = paste0(filepath,"output/mrna_rpf_fc_",comparisons[i],".png"), width=8,height=5) -->
<!-- } -->
<!-- ``` -->

<!-- ### GO and Regulatory enrichment analyses -->
<!-- ```{r} -->
<!-- source("~/R/my_scripts/human_go.R") -->
<!-- source("~/R/my_scripts/human_reg.R") -->

<!-- ctl_lin_rnadeg <- all_fc_data$Gene[all_fc_data$SigM & all_fc_data$comp=="ctl_lin" & abs(all_fc_data$FC_mRNA)>=1] -->
<!-- ctl_lin_rpfdeg <- all_fc_data$Gene[all_fc_data$SigR & all_fc_data$comp=="ctl_lin" & abs(all_fc_data$FC_RPF)>=1] -->

<!-- ctl_lin_go <- go_analyses(rna_genes = ctl_lin_rnadeg, rpf_genes = ctl_lin_rpfdeg) -->
<!-- ctl_lin_enrich <- reg_analyses(rna_genes = ctl_lin_rnadeg, rpf_genes = ctl_lin_rpfdeg) -->

<!-- ctl_imp_rnadeg <- all_fc_data$Gene[all_fc_data$SigM & all_fc_data$comp=="ctl_ctl_imp" & abs(all_fc_data$FC_mRNA)>=1] -->
<!-- ctl_imp_rpfdeg <- all_fc_data$Gene[all_fc_data$SigR & all_fc_data$comp=="ctl_ctl_imp" & abs(all_fc_data$FC_RPF)>=1] -->

<!-- ctl_ctl_imp_go <- go_analyses(rna_genes = ctl_imp_rnadeg, rpf_genes = ctl_imp_rpfdeg) -->
<!-- ctl_ctl_imp_enrich <- reg_analyses(rna_genes = ctl_imp_rnadeg, rpf_genes = ctl_imp_rpfdeg) -->

<!-- lin_lin_imp_rnadeg <- all_fc_data$Gene[all_fc_data$SigM & all_fc_data$comp=="lin_lin_imp" & abs(all_fc_data$FC_mRNA)>=1] -->
<!-- lin_lin_imp_rpfdeg <- all_fc_data$Gene[all_fc_data$SigR & all_fc_data$comp=="lin_lin_imp" & abs(all_fc_data$FC_RPF)>=1] -->

<!-- lin_lin_imp_go <- go_analyses(rna_genes = lin_lin_imp_rnadeg, rpf_genes = lin_lin_imp_rpfdeg) -->
<!-- lin_lin_imp_enrich <- reg_analyses(rna_genes = lin_lin_imp_rnadeg, rpf_genes = lin_lin_imp_rpfdeg) -->
<!-- ``` -->

<!-- # Plotting GO and reg analyses -->
<!-- ```{r} -->
<!-- # GO plots (custom) -->
<!-- source("~/R/my_scripts/ps_go_plot.R") -->
<!-- ctl_lin_go_plot <- ps_go_plot(ctl_lin_go,ontology = "all") -->
<!-- ctl_ctl_imp_go_plot <- ps_go_plot(ctl_ctl_imp_go,ontology = "all") -->
<!-- lin_lin_imp_go_plot <- ps_go_plot(lin_lin_imp_go,ontology = "all") -->

<!-- ggsave(ctl_lin_go_plot, filename = paste0(filepath,"output/ctl_lin_go.png"), width=8, height=5) -->
<!-- ggsave(ctl_ctl_imp_go_plot, filename = paste0(filepath,"output/ctl_ctl_imp_go.png"), width=8, height=5) -->
<!-- ggsave(lin_lin_imp_go_plot, filename = paste0(filepath,"output/lin_lin_imp_go.png"), width=8, height=5) -->

<!-- # Regulatory Enrichment plots (tRanslatome) -->
<!-- Radar(ctl_lin_enrich, n.nodes.1stlevel = 4, n.nodes.2ndlevel = 4) -->
<!-- Radar(ctl_ctl_imp_enrich, n.nodes.1stlevel = 4, n.nodes.2ndlevel = 4) -->
<!-- Radar(ctl_lin_enrich, n.nodes.1stlevel = 4, n.nodes.2ndlevel = 4) -->

<!-- ``` -->

<!-- ### Ratio of RPF and RNA fold changes -->
<!-- ```{r} -->
<!-- ggplot(all_fc_data) -->
<!-- ``` -->


<!-- ### DESeq2 analyses -->
<!-- ```{r} -->
<!-- a <- (rowMaxs(as.matrix(bcountData[,3:10]))>=64 & (rowMins(as.matrix(bcountData[,3:10]))>0) | rowMeans(as.matrix(bcountData[,3:10]))>64) -->
<!-- high_cov_data <- as.matrix(bcountData[a,3:10]) -->
<!-- rownames(high_cov_data) <- bcountData$gene[a] -->
<!-- high_cov_tpm <- as.matrix(btpmData[a,3:10]) -->
<!-- rownames(high_cov_tpm) <- bcountData$gene[a] -->

<!-- lin_deseq2 <- function(counts, cond, vcond){ -->
<!--   dds <- DESeqDataSetFromMatrix(counts, DataFrame(cond), ~ cond) -->
<!--   dds <- DESeq(dds) -->
<!--   res <- results(dds) -->
<!--   resLFC <- lfcShrink(dds, contrast=vcond, res=res) -->
<!-- } -->

<!-- comp <- c("ctl_fp","lin_fp") -->
<!-- ctl_lin_fp_lfc <- lin_deseq2(counts = high_cov_data[,comp], cond = factor(comp), vcond = c("cond",rev(comp))) -->
<!-- comp <- c("ctl_rn","lin_rn") -->
<!-- ctl_lin_rn_lfc <- lin_deseq2(counts = high_cov_data[,comp], cond = factor(comp), vcond = c("cond",rev(comp))) -->
<!-- comp <- c("ctl_fp","ctl_imp_fp") -->
<!-- ctl_ctl_imp_fp_lfc <- lin_deseq2(counts = high_cov_data[,comp], cond = factor(comp), vcond = c("cond",rev(comp))) -->
<!-- comp <- c("ctl_rn","ctl_imp_rn") -->
<!-- ctl_ctl_imp_rn_lfc <- lin_deseq2(counts = high_cov_data[,comp], cond = factor(comp), vcond = c("cond",rev(comp))) -->
<!-- comp <- c("lin_fp","lin_imp_fp") -->
<!-- lin_lin_imp_fp_lfc <- lin_deseq2(counts = high_cov_data[,comp], cond = factor(comp), vcond = c("cond",rev(comp))) -->
<!-- comp <- c("lin_rn","lin_imp_rn") -->
<!-- lin_lin_imp_rn_lfc <- lin_deseq2(counts = high_cov_data[,comp], cond = factor(comp), vcond = c("cond",rev(comp))) -->
<!-- ``` -->

<!-- ### Identify nature of transcriptional and translational reglulation. -->
<!-- ```{r} -->
<!-- # Read in differential mRNA/RPF data from DEGseq analyses. -->
<!-- comparisons <- c("ctl_lin","ctl_ctl_imp","lin_lin_imp") -->
<!-- comparisons_text <- c(expression(paste("wt vs ",Delta,italic("lin28b"))),  -->
<!--                       expression(paste("wt vs ",Delta,italic("imp1"))), -->
<!--                       expression(paste(Delta,italic("lin28b")," vs ",Delta,italic("lin28b"),Delta,italic("imp1")))) -->
<!-- all_comp <- list( -->
<!-- ctl_lin_fp_lfc, -->
<!-- ctl_lin_rn_lfc, -->
<!-- ctl_ctl_imp_fp_lfc, -->
<!-- ctl_ctl_imp_rn_lfc, -->
<!-- lin_lin_imp_fp_lfc, -->
<!-- lin_lin_imp_rn_lfc) -->

<!-- all_fc_data <- c() -->
<!-- for(i in 1:3){ -->
<!--   str_m <- data.frame(all_comp[[2*i]][,c("log2FoldChange","padj")]) -->
<!--   str_r <- data.frame(all_comp[[2*i-1]][,c("log2FoldChange","padj")]) -->
<!--   str_m <- cbind(rownames(str_m),str_m) -->
<!--   str_r <- cbind(rownames(str_r),str_r) -->

<!--   names(str_m) <- c("Gene","FC_mRNA","SigM") -->
<!--   names(str_r) <- c("Gene","FC_RPF","SigR") -->
<!--   str_all <- full_join(str_m,str_r,by="Gene") -->
<!--   str_all$comp <- comparisons[i] -->
<!--   all_fc_data <- bind_rows(all_fc_data,str_all) -->
<!-- } -->
<!-- all_fc_data$SigM <- (all_fc_data$SigM<0.05) -->
<!-- all_fc_data$SigR <- (all_fc_data$SigR<0.05) -->

<!-- # Identify change categories -->
<!-- all_fc_data$Change <- NA -->
<!-- all_fc_data$Change[abs(all_fc_data$FC_mRNA)>1 & abs(all_fc_data$FC_RPF-all_fc_data$FC_mRNA)<=1] <- "Transcription only" -->
<!-- all_fc_data$Change[abs(all_fc_data$FC_RPF)>1 & abs(all_fc_data$FC_mRNA)<=1] <- "Translation only" -->
<!-- all_fc_data$Change[abs(all_fc_data$FC_mRNA)>1 & abs(all_fc_data$FC_RPF-all_fc_data$FC_mRNA)>1] <- "Translation reinforcement" -->
<!-- all_fc_data$Change[abs(all_fc_data$FC_mRNA)>1 & abs(all_fc_data$FC_RPF)>1 & sign(all_fc_data$FC_mRNA)==sign(all_fc_data$FC_RPF) & abs(all_fc_data$FC_mRNA-all_fc_data$FC_RPF)>1 & abs(all_fc_data$FC_mRNA)>abs(all_fc_data$FC_RPF)] <- "Translation buffering (partial)" -->
<!-- all_fc_data$Change[abs(all_fc_data$FC_mRNA)>1 & abs(all_fc_data$FC_RPF)<=1] <- "Translation buffering (complete)" -->
<!-- all_fc_data$Change[abs(all_fc_data$FC_mRNA)>1 & abs(all_fc_data$FC_RPF)>1 & sign(all_fc_data$FC_mRNA)!=sign(all_fc_data$FC_RPF)] <- "Translation antagonism" -->
<!-- all_fc_data$Change[abs(all_fc_data$FC_mRNA)<=1 & abs(all_fc_data$FC_RPF)<=1] <- "No change" -->
<!-- all_fc_data$Change <- as.factor(all_fc_data$Change) -->
<!-- all_fc_data$Change <- factor(all_fc_data$Change,levels(all_fc_data$Change)[c(1,2,6,7,5,4,3)]) -->

<!-- all_fc_data <- all_fc_data[!is.na(all_fc_data$Change),] -->

<!-- write.table(all_fc_data,paste0(filepath,"all_mrna_rpf_foldchange.tsv"),quote=F,sep="\t",row=F,col=T) -->

<!-- for(i in 1:3){ -->
<!--   m <- lm(all_fc_data$FC_RPF[all_fc_data$comp==comparisons[i]] ~ all_fc_data$FC_mRNA[all_fc_data$comp==comparisons[i]]) -->
<!--   mrna_rpf_fc_change <- ggplot(filter(all_fc_data,comp==comparisons[i]), aes(x=FC_mRNA, y=FC_RPF, col=Change)) + -->
<!--     geom_point(alpha=0.5) + -->
<!--     scale_alpha(guide = 'none') +  -->
<!--     scale_color_brewer(palette = "Set1") + -->
<!--     theme(legend.title=element_blank()) +  -->
<!--     labs(x="mRNA log2 fold-change", y= "RPF log2 fold-change") +  -->
<!--     guides(colour = guide_legend(override.aes = list(size=5,aplha=0.8))) + -->
<!--     xlim(-2,2) + -->
<!--     ylim(-2,2) + -->
<!--     coord_fixed(ratio = 1) + -->
<!--     annotate("text", x = -1.5, y = 1.75, label = comparisons_text[i]) + -->
<!--     annotate("text", x = -1.5, y = 1.5, label = r2plot(m), parse=TRUE) -->

<!--   mrna_rpf_fc_change -->
<!--   ggsave(mrna_rpf_fc_change, filename = paste0(filepath,"plots/mrna_rpf_fc_",comparisons[i],".png"), width=8,height=5) -->
<!-- } -->
<!-- ``` -->

<!-- ### Riborex -->
<!-- ```{r} -->
<!-- cond <- factor(rep(c("Control","LIN"),each=1)) -->
<!-- ctl_lin_rx_te <- riborex(rnaCntTable = high_cov_data[,c("ctl_rn","lin_rn")], riboCntTable = high_cov_data[,c("ctl_fp","lin_fp")], riboCond = cond, rnaCond = cond) -->

<!-- cond <- factor(rep(c("Control","Control-IMP"),each=1)) -->
<!-- ctl_ctl_imp_rx_te <- riborex(rnaCntTable = high_cov_data[,c("ctl_rn","ctl_imp_rn")], riboCntTable = high_cov_data[,c("ctl_fp","ctl_imp_fp")], riboCond = cond, rnaCond = cond) -->

<!-- cond <- factor(rep(c("LIN","LIN-IMP"),each=1)) -->
<!-- lin_lin_im_rx_te <- riborex(rnaCntTable = high_cov_data[,c("lin_rn","lin_imp_rn")], riboCntTable = high_cov_data[,c("lin_fp","lin_imp_fp")], riboCond = cond, rnaCond = cond) -->

<!-- # teData[rownames(res.deseq2),"padj"] <- res.deseq2$padj -->
<!-- # write.table(teData,paste0(filepath,"all_TE.tsv"),quote=F,sep="\t",row=T,col=T) -->
<!-- ``` -->

<!-- ### Read and process eclip data -->
<!-- ```{r} -->
<!-- eclip <- read.csv("/data/riboseq/Rustgi/eClip.txt",h=T,sep="\t",stringsAsFactors = F) -->
<!-- colnames(eclip)[1] <- "gene" -->
<!-- eclip_proc <- eclip[,1:2] -->
<!-- eclip_proc$IMP1_CDS <- rowMeans(eclip[,c("IMP1_Rep1_CDS","IMP1_Rep2_CDS")],na.rm=TRUE) -->
<!-- eclip_proc$IMP1_UTR3 <- rowMeans(eclip[,c("IMP1_Rep1_3_UTR","IMP1_Rep2_3_UTR")],na.rm=TRUE) -->
<!-- eclip_proc$IMP2_CDS <- rowMeans(eclip[,c("IMP2_Rep1_CDS","IMP2_Rep2_CDS")],na.rm=TRUE) -->
<!-- eclip_proc$IMP2_UTR3 <- rowMeans(eclip[,c("IMP2_Rep1_3_UTR","IMP2_Rep2_3_UTR")],na.rm=TRUE) -->
<!-- eclip_proc$IMP3_CDS <- eclip[,c("IMP3_CDS")] -->
<!-- eclip_proc$IMP3_UTR3 <- eclip[,c("IMP3_3_UTR")] -->
<!-- eclip_proc <- eclip_proc[!duplicated(eclip_proc$gene),] -->
<!-- ``` -->

<!-- ## TE analyses -->
<!-- ```{r} -->
<!-- ctl_te <- high_cov_tpm[,"ctl_fp"]/high_cov_tpm[,"ctl_rn"] -->
<!-- lin_te <- high_cov_tpm[,"lin_fp"]/high_cov_tpm[,"lin_rn"] -->
<!-- ctl_imp_te <- high_cov_tpm[,"ctl_imp_fp"]/high_cov_tpm[,"ctl_imp_rn"] -->
<!-- lin_imp_te <- high_cov_tpm[,"lin_imp_fp"]/high_cov_tpm[,"lin_imp_rn"] -->

<!-- teData <- data.frame(ctl_te,lin_te,ctl_imp_te,lin_imp_te) -->
<!-- teData$gene <- rownames(teData) -->

<!-- teData_clip <- left_join(teData,eclip_proc,"gene") -->

<!-- teScatPlot <- ggplot(teData_clip%>%filter(!is.na(IMP1_UTR3)), aes(ctl_te, lin_te)) + -->
<!-- # tePlot <- ggplot(teData_clip, aes(ctl_te, ctl_imp_te, col=IMP1_UTR3>1)) + -->
<!--   scale_x_log10(limits=c(1e-3,1e2)) +  -->
<!--   scale_y_log10(limits=c(1e-3,1e2)) +  -->
<!--   # geom_point() + -->
<!--   geom_point(aes(col=IMP1_UTR3)) + -->
<!--   coord_equal(ratio = 1) + -->
<!--   geom_abline(slope = 1,intercept = 0) + -->
<!--   # scale_color_gradient2(low="black",mid="white",high="yellow",na.value = "gray",midpoint = 1.6, limits=c(-1,4)) + -->
<!--   scale_color_gradient2(low="#e41a1c",mid="white",high="#377eb8",na.value = "grey60") + -->
<!--   # scale_color_gradient2(low="#e41a1c",mid="white",high="#377eb8",na.value = "grey60",midpoint = quantile(teData_clip$IMP1_UTR3,0.5,na.rm=T)) -->
<!-- # , limits=c(quantile(teData_clip$IMP1_UTR3,0.2,na.rm=T),quantile(teData_clip$IMP1_UTR3,0.8,na.rm=T))) + -->
<!--   labs(x="WT Translation Efficiency (TE)", y="LIN Translation Efficiency (TE)")  -->
<!--   # theme_classic() -->

<!-- # tePlot -->
<!-- teScatPlot -->
<!-- ggsave(teScatPlot, filename = paste0(filepath,"output/ctl_lin_te_eclip.png"), width=7, height=5) -->


<!-- teBoxdata <- teData_clip -->
<!-- teBoxdata$category <- NA -->
<!-- teBoxdata$category[teBoxdata$IMP1_UTR3<= quantile(teData_clip$IMP1_UTR3,0.2,na.rm=T)] <- "Very low" -->

<!-- teBoxdata$category[teBoxdata$IMP1_UTR3<= quantile(teData_clip$IMP1_UTR3,0.4,na.rm=T) & teBoxdata$IMP1_UTR3 > quantile(teData_clip$IMP1_UTR3,0.2,na.rm=T)] <- "Low" -->

<!-- teBoxdata$category[teBoxdata$IMP1_UTR3<= quantile(teData_clip$IMP1_UTR3,0.6,na.rm=T) & teBoxdata$IMP1_UTR3 > quantile(teData_clip$IMP1_UTR3,0.4,na.rm=T)] <- "Medium" -->

<!-- teBoxdata$category[teBoxdata$IMP1_UTR3<= quantile(teData_clip$IMP1_UTR3,0.8,na.rm=T) & teBoxdata$IMP1_UTR3 > quantile(teData_clip$IMP1_UTR3,0.6,na.rm=T)] <- "High" -->

<!-- teBoxdata$category[teBoxdata$IMP1_UTR3 > quantile(teData_clip$IMP1_UTR3,0.8,na.rm=T)] <- "Very high" -->

<!-- teBoxdata$category <- factor(teBoxdata$category, levels = c("Very low", "Low", "Medium", "High", "Very high")) -->

<!-- teBoxplot <- ggplot(teBoxdata, aes(x=category, y=log2(ctl_imp_te/ctl_te), fill=category)) + -->
<!--   geom_violin() + -->
<!--   labs(x="eClip binding", y="Log2 WT-IMP / WT Translation Efficiency (TE)") + -->
<!--   guides(fill=FALSE) -->
<!-- teBoxplot -->
<!-- ggsave(teBoxplot, filename = paste0(filepath,"output/ctl_ctl_imp_te_eclip_box.png"), width=5,height=5) -->

<!-- ``` -->

<!-- ### Position-specific plots for WNT genes -->
<!-- ```{r} -->
<!-- # RAC1, APC, CTNNB1, and BTRC -->
<!-- wnt <- c("RAC1", "APC", "CTNNB1", "BTRC") -->
<!-- names(wnt) <- c("ENST00000348035","ENST00000257430","ENST00000405570","ENST00000370187") -->

<!-- fid1 <- H5Fopen("/data/riboseq/Rustgi/ctl_fp/ctl_fp.h5") -->
<!-- fid2 <- H5Fopen("/data/riboseq/Rustgi/lin_fp/lin_fp.h5") -->
<!-- fid3 <- H5Fopen("/data/riboseq/Rustgi/lin_imp_fp/lin_imp_fp.h5") -->

<!-- allout <- c() -->

<!-- for(gene in names(wnt)){ -->
<!--   a1 <- colSums(H5Dread(H5Dopen(fid1,paste0("/",gene,"/","data","/reads/data")))) -->
<!--   b1 <- a1/mean(a1) -->
<!--   tmp1 <- cbind(as.character(1:length(a1)),wnt[gene],as.character(a1),as.character(b1),"WT") -->

<!--   a2 <- colSums(H5Dread(H5Dopen(fid2,paste0("/",gene,"/","data","/reads/data")))) -->
<!--   b2 <- a2/mean(a2) -->
<!--   tmp2 <- cbind(as.character(1:length(a2)),wnt[gene],as.character(a2),as.character(b2),"LIN") -->

<!--   a3 <- colSums(H5Dread(H5Dopen(fid3,paste0("/",gene,"/","data","/reads/data")))) -->
<!--   b3 <- a3/mean(a3) -->
<!--   tmp3 <- cbind(as.character(1:length(a3)),wnt[gene],as.character(a3),as.character(b3),"LIN-IMP") -->

<!--   allout <- rbind(allout,tmp1,tmp2,tmp3) -->
<!-- } -->
<!-- H5close() -->

<!-- # Data frame -->
<!-- allout <- as.data.frame(allout) -->
<!-- colnames(allout) <- c("Position","Gene","Count","NormCount","Data") -->
<!-- allout$Position <- as.integer(as.character(allout$Position)) -->
<!-- allout$Count <- as.integer(as.character(allout$Count)) -->
<!-- allout$NormCount <- as.numeric(as.character(allout$NormCount)) -->
<!-- allout$Data <- factor(allout$Data, levels=c("WT","LIN","LIN-IMP")) -->

<!-- # Plots -->
<!-- # posplot_all <- ggplot(allout, aes(x=Position, y=NormCount, col=Data)) + -->
<!-- #   geom_line(alpha=0.5) + -->
<!-- #   facet_wrap(~Gene, scales="free") + -->
<!-- #   scale_color_brewer(palette = "Set1") +  -->
<!-- #   theme_bw() +  -->
<!-- #   theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) -->

<!-- posplot_all <- ggplot(allout, aes(x=Position, y=NormCount, col=Data, fill=Data)) + -->
<!--   geom_bar(stat='identity',position="dodge") + -->
<!--   facet_wrap(~Gene, scales="free") + -->
<!--   scale_color_brewer(palette = "Set1") +  -->
<!--   theme_bw() +  -->
<!--   theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) -->
<!-- ggsave(posplot_all, filename = paste0(filepath,"output/pos_sp_plot_all.png"), width=10,height=8) -->
<!-- ``` -->

<!-- # Position plot for only CDS -->
<!-- ```{r} -->
<!-- wnt <- c("RAC1", "APC", "CTNNB1", "BTRC") -->
<!-- names(wnt) <- c("ENST00000348035","ENST00000257430","ENST00000405570","ENST00000370187") -->

<!-- fid1 <- H5Fopen("/data/riboseq/Rustgi/ctl_fp/ctl_fp.h5") -->
<!-- fid2 <- H5Fopen("/data/riboseq/Rustgi/lin_fp/lin_fp.h5") -->
<!-- fid3 <- H5Fopen("/data/riboseq/Rustgi/lin_imp_fp/lin_imp_fp.h5") -->

<!-- allout2 <- c() -->

<!-- for(gene in names(wnt)){ -->
<!--   a1 <- colSums(H5Dread(H5Dopen(fid1,paste0("/",gene,"/","data","/reads/data")))) -->
<!--   a1 <- a1[start(gff[gff$Name==gene & gff$type=="CDS",]):end(gff[gff$Name==gene & gff$type=="CDS",])] -->
<!--   b1 <- a1/mean(a1) -->
<!--   tmp1 <- cbind(as.character(1:length(a1)),wnt[gene],as.character(a1),as.character(b1),"WT") -->

<!--   a2 <- colSums(H5Dread(H5Dopen(fid2,paste0("/",gene,"/","data","/reads/data")))) -->
<!--   a2 <- a2[start(gff[gff$Name==gene & gff$type=="CDS",]):end(gff[gff$Name==gene & gff$type=="CDS",])] -->
<!--   b2 <- a2/mean(a2) -->
<!--   tmp2 <- cbind(as.character(1:length(a2)),wnt[gene],as.character(a2),as.character(b2),"LIN") -->

<!--   a3 <- colSums(H5Dread(H5Dopen(fid3,paste0("/",gene,"/","data","/reads/data")))) -->
<!--   a3 <- a3[start(gff[gff$Name==gene & gff$type=="CDS",]):end(gff[gff$Name==gene & gff$type=="CDS",])] -->
<!--   b3 <- a3/mean(a3) -->
<!--   tmp3 <- cbind(as.character(1:length(a3)),wnt[gene],as.character(a3),as.character(b3),"LIN-IMP") -->

<!--   allout2 <- rbind(allout2,tmp1,tmp2,tmp3) -->
<!-- } -->
<!-- H5close() -->

<!-- # Data frame -->
<!-- allout2 <- as.data.frame(allout2) -->
<!-- colnames(allout2) <- c("Position","Gene","Count","NormCount","Data") -->
<!-- allout2$Position <- as.integer(as.character(allout2$Position)) -->
<!-- allout2$Count <- as.integer(as.character(allout2$Count)) -->
<!-- allout2$NormCount <- as.numeric(as.character(allout2$NormCount)) -->
<!-- allout2$Data <- factor(allout2$Data, levels=c("WT","LIN","LIN-IMP")) -->

<!-- # Plots -->
<!-- posplot_cds <- ggplot(allout2, aes(x=Position, y=NormCount, col=Data, fill=Data)) + -->
<!--   geom_bar(stat='identity',position="dodge") + -->
<!--   facet_wrap(~Gene, scales="free") + -->
<!--   scale_color_brewer(palette = "Set1") +  -->
<!--   theme_bw() +  -->
<!--   theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) -->
<!-- ggsave(posplot_cds, filename = paste0(filepath,"output/pos_sp_plot_cds.png"), width=10,height=8) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- a <- ggplot(filter(allout2,Gene=="BTRC"), aes(x=Position, y=NormCount, col=Data)) + -->
<!--   geom_line() + -->
<!--   facet_wrap(~Data,nrow=3) + -->
<!--   scale_color_brewer(palette = "Set1") +  -->
<!--   theme_bw() +  -->
<!--   theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) -->
<!-- a -->
<!-- ggsave(a, filename = paste0(filepath,"output/pos_sp_plot_cds_BTRC.png"), width=10,height=8) -->
<!-- ``` -->


<!-- ## TE analyses all -->
<!-- ```{r} -->
<!-- ctl_te <- btpmData[,"ctl_fp"]/btpmData[,"ctl_rn"] -->
<!-- lin_te <- btpmData[,"lin_fp"]/btpmData[,"lin_rn"] -->
<!-- ctl_imp_te <- btpmData[,"ctl_imp_fp"]/btpmData[,"ctl_imp_rn"] -->
<!-- lin_imp_te <- btpmData[,"lin_imp_fp"]/btpmData[,"lin_imp_rn"] -->

<!-- # ctl_te <- btpmData[,"ctl_fp"] -->
<!-- # lin_te <- btpmData[,"lin_fp"] -->
<!-- # ctl_imp_te <- btpmData[,"ctl_imp_fp"] -->
<!-- # lin_imp_te <- btpmData[,"lin_imp_fp"] -->

<!-- teData_all <- data.frame(ctl_te,lin_te,ctl_imp_te,lin_imp_te) -->
<!-- teData_all$gene <- as.character(btpmData$gene) -->

<!-- teData_clip_all <- left_join(teData_all,eclip_proc,"gene") -->

<!-- teScatPlot <- ggplot(teData_clip_all %>% filter(!is.na(IMP1_UTR3)), aes(ctl_te, lin_te)) +  -->
<!-- # tePlot <- ggplot(teData_clip, aes(ctl_te, ctl_imp_te, col=IMP1_UTR3>1)) +  -->
<!--   scale_x_log10(limits=c(1e-3,1e2)) +  -->
<!--   scale_y_log10(limits=c(1e-3,1e2)) +  -->
<!--   # geom_point() +  -->
<!--   geom_point(aes(color=IMP1_UTR3)) + -->
<!--   coord_equal(ratio = 1) + -->
<!--   geom_abline(slope = 1,intercept = 0) + -->
<!--   scale_colour_gradient2() + -->
<!--   labs(x="WT Translation Efficiency (TE)", y="LIN Translation Efficiency (TE)") -->

<!-- teScatPlot -->

<!-- a <- teData_clip_all -->
<!-- a[,1:4] <- a[,1:4]/a[,1] -->
<!-- teBoxdata <- a %>%  -->
<!--                   filter(!is.na(IMP1_UTR3)) %>% -->
<!--                   select(ctl_te, lin_te,ctl_imp_te, lin_imp_te,IMP1_UTR3) %>% -->
<!--                   gather(class,TE,1:4) %>% -->
<!--                   filter(!is.na(TE) & is.finite(TE)) -->

<!-- teBoxdata$category <- NA -->
<!-- teBoxdata$category[teBoxdata$IMP1_UTR3<= -2] <- "Very low" -->
<!-- teBoxdata$category[teBoxdata$IMP1_UTR3<= -0.5 & teBoxdata$IMP1_UTR3 >-2] <- "Low" -->
<!-- teBoxdata$category[teBoxdata$IMP1_UTR3<= 0.5 & teBoxdata$IMP1_UTR3 >-0.5] <- "Medium" -->
<!-- teBoxdata$category[teBoxdata$IMP1_UTR3<= 2 & teBoxdata$IMP1_UTR3 > 0.5] <- "High" -->
<!-- teBoxdata$category[teBoxdata$IMP1_UTR3 > 2] <- "Very high" -->
<!-- teBoxdata$category <- factor(teBoxdata$category, levels = c("Very low", "Low", "Medium", "High", "Very high")) -->

<!-- # teBoxdata$category[teBoxdata$IMP1_UTR3 < -1] <- "Low" -->
<!-- # teBoxdata$category[teBoxdata$IMP1_UTR3 > 1] <- "High" -->
<!-- # teBoxdata$category <- factor(teBoxdata$category, levels = c("Low","High")) -->

<!-- teBox <- ggplot(teBoxdata, aes(x=class,y=TE,col=category)) +  -->
<!--   scale_y_log10() + -->
<!--   geom_boxplot() -->
<!-- teBox -->

<!-- ``` -->



