---
title: "RNA-Seq and Ribo-Seq TPMs of Dengue"
output: html_document
---

```{r Packages Needed, message=FALSE, warning=FALSE}
library(rhdf5)
library(tidyverse)
library(Rmisc)
```

```{r rhdf5 setup}
# Accessing H5 files
fid <- "/data/riboseq/Dengue/output/"

file <- c("RiboNSC1","RiboNSC2","RiboNSC3","RibosiP1","RibosiP2","RibosiP3","RNANSC1","RNANSC2","RNANSC3","RNAsiP1","RNAsiP2","RNAsiP3") # Names of h5 files

# Dengue Gene Positions
dengue_gene_pos <- as.data.frame(read_delim("/data/riboseq/Dengue/output/dengue.gene_pos", 
    "\t", escape_double = FALSE, col_names = FALSE, 
    trim_ws = TRUE))
dengue_gene_pos$gene_length <- (dengue_gene_pos[,3] - dengue_gene_pos[,2]) + 1 #Gene lengths
dengue_gene_pos_matrix <- as.matrix(dengue_gene_pos[,2:4])
rownames(dengue_gene_pos_matrix) <- dengue_gene_pos[,1]
colnames(dengue_gene_pos_matrix) <-c("start","end","gene_length")
dengue_gene_pos <- dengue_gene_pos_matrix
```

```{r All Samples Counts and RPKMs}
# All Counts for each sample
all_counts <- read_delim("/data/riboseq/Dengue/output/all_counts_bam.tsv", 
    "\t", escape_double = FALSE, trim_ws = TRUE)
all_counts <- as.data.frame(all_counts)

# All RPKMs for each sample
all_rpkms <- read_delim("/data/riboseq/Dengue/output/all_rpkms_bam.tsv", 
    "\t", escape_double = FALSE, trim_ws = TRUE)
all_rpkms <- as.data.frame(all_rpkms)
```

```{r RPF Reads for Lengths 29-35}
RPF_files <- file[1:6] # Isolating only the Ribo-seq

# Creating a blank matrix for data input
reads <- matrix(ncol = length(RPF_files),nrow = nrow(dengue_gene_pos))
colnames(reads) <- RPF_files
rownames(reads) <- rownames(dengue_gene_pos)

for(x in 1:length(RPF_files)){
filename <- RPF_files[x]  
h5id <- H5Fopen(paste0(fid,filename,".h5"))
matrix <- H5Dread(H5Dopen(h5id,"/DENGUE/data/reads/data"))

# UTR5 Read Count
    limits <- dengue_gene_pos[1,1:2]
    
    l29_30_5 <- sum(matrix[20:21,(limits[1]):(limits[2]-15)]) #Reads of lengths 29-30, offset by 15 nts
    l31_33_5 <- sum(matrix[22:24,(limits[1]):(limits[2]-16)]) #Reads between lengths 31-33, offset by 16 nts
    l34_35_5 <- sum(matrix[25:26,(limits[1]):(limits[2]-17)]) #Reads of lengths 34-35, offset by 17 nts

    reads[1,x] <- sum(l29_30_5,l31_33_5,l34_35_5)
    
# UTR3 Read Count
    limits <- dengue_gene_pos[nrow(dengue_gene_pos),1:2]
    l29_30_3 <- sum(matrix[20:21,(limits[1]-15):(limits[2])]) #Reads of lengths 29-30, offset by 15 nts
    l31_33_3 <- sum(matrix[22:24,(limits[1]-16):(limits[2])]) #Reads between lengths 31-33, offset by 16 nts
    l34_35_3 <- sum(matrix[25:26,(limits[1]-17):(limits[2])]) #Reads of lengths 34-35, offset by 17 nts

    reads[nrow(dengue_gene_pos),x] <- sum(l29_30_3,l31_33_3,l34_35_3)    
    
# All Other Genes' Read Counts  
  for(i in 2:(nrow(dengue_gene_pos)-1)){
    limits <- dengue_gene_pos[i,1:2]
    l29_30_ao <- sum(matrix[20:21,(limits[1]-15):(limits[2]-15)]) #Reads of lengths 29-30, offset by 15 nts
    l31_33_ao <- sum(matrix[22:24,(limits[1]-16):(limits[2]-16)]) #Reads between lengths 31-33, offset by 16 nts
    l34_35_ao <- sum(matrix[25:26,(limits[1]-17):(limits[2]-17)]) #Reads between lengths 34-35, offset by 17 nts
    
      reads[i,x] <- sum(l29_30_ao,l31_33_ao,l34_35_ao)
  }
H5Fclose(h5id)
}
```

```{r RNA reads 50 nts}
# Mapping counts to Dengue genes in RNA-seqs
RNA_files <- file[7:12] # Isolating only the RNA-seq

# Creating a blank matrix for data input
reads2 <-matrix(ncol = length(RNA_files),nrow = nrow(dengue_gene_pos))
colnames(reads2) <- RNA_files
rownames(reads2) <- rownames(dengue_gene_pos)

for(x in 1:length(RNA_files)){
filename <- RNA_files[x]  
h5id <- H5Fopen(paste0(fid,filename,".h5"))
matrix <- H5Dread(H5Dopen(h5id,"/DENGUE/data/reads/data"))

# UTR5 Read Count
    limits <- dengue_gene_pos[1,1:2]
    ln5 <- sum(matrix[,(limits[1]):(limits[2]-50)]) # Reads with gene boundary offset by 50 nts

    reads2[1,x] <- ln5
    
# UTR3 Read Count
    limits <- dengue_gene_pos[nrow(dengue_gene_pos),1:2]
    ln3 <- sum(matrix[,(limits[1]-50):(limits[2])]) # Reads with gene boundary offset by 50 nts

    reads2[nrow(dengue_gene_pos),x] <- ln3
    
# All Other Genes' Read Counts  
  for(i in 2:(nrow(dengue_gene_pos)-1)){
    limits <- dengue_gene_pos[i,1:2]
    lnao <- sum(matrix[,(limits[1]-50):(limits[2]-50)]) #Reads with gene boundary offset by 50 nts at both ends
      reads2[i,x] <- lnao
  }
}
```

```{r RNA reads 25 nts}
# Mapping counts to Dengue genes in RNA-seqs

# Creating a blank matrix for data input
reads3 <-matrix(ncol = length(RNA_files),nrow = nrow(dengue_gene_pos))
colnames(reads3) <- RNA_files
rownames(reads3) <- rownames(dengue_gene_pos)

for(x in 1:length(RNA_files)){
filename <- RNA_files[x]  
h5id <- H5Fopen(paste0(fid,filename,".h5"))
matrix <- H5Dread(H5Dopen(h5id,"/DENGUE/data/reads/data"))

# UTR5 Read Count
    limits <- dengue_gene_pos[1,1:2]
    ln5 <- sum(matrix[,(limits[1]):(limits[2]-25)]) #Reads with gene boundary offset by 25 nts

    reads3[1,x] <- ln5
    
# UTR3 Read Count
    limits <- dengue_gene_pos[nrow(dengue_gene_pos),1:2]
    ln3 <- sum(matrix[,(limits[1]-25):(limits[2])]) #Reads with gene boundary offset by 25 nts

    reads3[nrow(dengue_gene_pos),x] <- ln3
    
# All Other Genes' Read Counts  
  for(i in 2:(nrow(dengue_gene_pos)-1)){
    limits <- dengue_gene_pos[i,1:2]
    lnao <- sum(matrix[,(limits[1]-25):(limits[2]-25)]) #Reads with gene boundary offset by 25 nts at both ends
      reads3[i,x] <- lnao
  }
}
```

```{r Calculating RPKM TPM}
total_sample_counts <- colSums(all_counts[,3:14]) # Sum of all the counts for each sample

# Dengue Ribo-seq RPKM
Dengue_Ribo_RPKM <- t(t(reads)/(total_sample_counts[1:6]*all_counts[all_counts$tid == "DENGUE","Length"]))*10^9

# Dengue RNA-seq RPKM
Dengue_RNA_RPKM_50 <- t(t(reads2)/(total_sample_counts[7:12]*all_counts[all_counts$tid == "DENGUE","Length"]))*10^9 # For the 50 nt offset

Dengue_RNA_RPKM_25 <- t(t(reads3)/(total_sample_counts[7:12]*all_counts[all_counts$tid == "DENGUE","Length"]))*10^9 # For the 25 nt offset

total_sample_RPKM <- colSums(all_rpkms[,3:14]) # Sum of all the RPKMs for each sample

# Dengue Ribo-seq & RNA-seq TPM
Dengue_Ribo_TPM <- (Dengue_Ribo_RPKM/total_sample_RPKM[1:6])*10^6
Dengue_RNA_TPM_50 <- (Dengue_RNA_RPKM_50/total_sample_RPKM[7:12])*10^6
Dengue_RNA_TPM_25 <- (Dengue_RNA_RPKM_25/total_sample_RPKM[7:12])*10^6

```

```{r}
# Preparing Dataframes for Graphing
Dengue_Ribo_TPM_DF <- as.data.frame(Dengue_Ribo_TPM)
Dengue_RNA_TPM_50_DF <- as.data.frame(Dengue_RNA_TPM_50)
Dengue_RNA_TPM_25_DF <- as.data.frame(Dengue_RNA_TPM_25)

Dengue_Ribo_TPM_DF$Control <- rowMeans(Dengue_Ribo_TPM_DF[,1:3]) # Average of the controls
Dengue_Ribo_TPM_DF$Knockdown <- rowMeans(Dengue_Ribo_TPM_DF[,4:6]) # Average of knockdown

Dengue_RNA_TPM_50_DF$Control <- rowMeans(Dengue_RNA_TPM_50_DF[,1:3]) # Average of the controls
Dengue_RNA_TPM_50_DF$Knockdown <- rowMeans(Dengue_RNA_TPM_50_DF[,4:6]) # Average of knockdown

Dengue_RNA_TPM_25_DF$Control <- rowMeans(Dengue_RNA_TPM_25_DF[,1:3]) # Average of the controls
Dengue_RNA_TPM_25_DF$Knockdown <- rowMeans(Dengue_RNA_TPM_25_DF[,4:6]) # Average of knockdown

# Reshaping dataframes for graphing 
test1 <- Dengue_Ribo_TPM_DF[,7:8] %>% gather(key = sample, value = Ribo_tpm)
test2 <- Dengue_RNA_TPM_50_DF[,7:8] %>% gather(key = sample, value = RNA_50_tpm)
test3 <- Dengue_RNA_TPM_25_DF[,7:8] %>% gather(key = sample, value = RNA_25_tpm)
test <- cbind.data.frame(test1,test2,test3)

# Creating TPM_RPF/TPM_RNA ratio for control vs. knockdown
test_rpf_rna50 <- cbind.data.frame(sample = test[,1],tpm_ratio = test[,2]/test[,4])
test_rpf_rna50[is.na(test_rpf_rna50[,2]),2] <- 0 # Replacing NaNs with 0
test_rpf_rna50$gene <- factor(rep(rownames(dengue_gene_pos),2),levels = rownames(dengue_gene_pos))

test_rpf_rna25 <- cbind.data.frame(sample = test[,1],tpm_ratio = test[,2]/test[,6])
test_rpf_rna25[is.na(test_rpf_rna25[,2]),2] <- 0
test_rpf_rna25$gene <- factor(rep(rownames(dengue_gene_pos),2),levels = rownames(dengue_gene_pos))
```

```{r}
# Plotting TPM_RPF/TPM_RNA ratio for control vs. knockdown for 50 nts offset for RNA-seq
ggplot(test_rpf_rna50,aes(x=gene,y=tpm_ratio,fill=sample)) + geom_col(position=position_dodge(width=0.9)) + scale_y_continuous("TPM_RPF/TPM_RNA (50 nts offset)",limits = c(0,0.15)) + labs(title = "TPM_RPF/TPM_RNA ratio for control vs. knockdown ordered by Dengue genes")
```

```{r TPM_RPF/TPM_RNA Graph 2}
# Plotting TPM_RPF/TPM_RNA ratio for control vs. knockdown for 25 nts offset for RNA-seq
ggplot(test_rpf_rna25,aes(x=gene,y=tpm_ratio,fill=sample)) + geom_col(position=position_dodge(width=0.9)) + scale_y_continuous("TPM_RPF/TPM_RNA (25 nts offset)",limits = c(0,0.15)) + labs(title = "TPM_RPF/TPM_RNA ratio for control vs. knockdown ordered by Dengue genes")
```

```{r}
Dengue_TE_50 <- Dengue_Ribo_TPM_DF[,1:6]/ Dengue_RNA_TPM_50_DF[,1:6]
Dengue_TE_25 <- Dengue_Ribo_TPM_DF[,1:6]/ Dengue_RNA_TPM_25_DF[,1:6]
colnames(Dengue_TE_50) <- c("TENSC1","TENSC2","TENSC3","TEsiP1","TEsiP2","TEsiP3")
colnames(Dengue_TE_25) <- c("TENSC1","TENSC2","TENSC3","TEsiP1","TEsiP2","TEsiP3")
Dengue_TE_50$Gene <- rownames(Dengue_TE_50)
Dengue_TE_25$Gene <- rownames(Dengue_TE_25)
Dengue_TE_50$Gene <- factor(Dengue_TE_50$Gene,levels=Dengue_TE_50$Gene[1:13])
Dengue_TE_25$Gene <- factor(Dengue_TE_25$Gene,levels=Dengue_TE_25$Gene[1:13])

Dengue_TE_50 <- Dengue_TE_50 %>% gather(dtype, TE, 1:6)
Dengue_TE_25 <- Dengue_TE_25 %>% gather(dtype, TE, 1:6)

Dengue_TE_50$Dataset <- substr(Dengue_TE_50$dtype,3,5)
Dengue_TE_25$Dataset <- substr(Dengue_TE_25$dtype,3,5)

D_te25_se <- summarySE(Dengue_TE_25, measurevar = "TE", groupvars = c("Gene","Dataset"))

te_plot <- ggplot(D_te25_se, aes(x=Gene, y=TE, fill=Dataset)) + 
  geom_bar(position=position_dodge(), stat="identity") +
  geom_errorbar(aes(ymin=TE-se, ymax=TE+se),
                width=.2,                    # Width of the error bars
                position=position_dodge(.9)) + 
  labs(y="Translation Efficiency (RPF TPM/mRNA TPM)")

te_plot

ggsave(te_plot, filename = paste0(filepath,"dengue_te_plot.png"), width=8, height=5)

tmpdat <- D_te25_se
tmpdat[,4:7] <- round(tmpdat[,4:7], 4)
write_tsv(tmpdat, path=paste0(filepath,"dengue_te.tsv"))

Dengue_Ribo_TPM <- round(Dengue_Ribo_TPM, 4)
tmpdat <- as.data.frame(cbind(rownames(Dengue_Ribo_TPM), Dengue_Ribo_TPM))
colnames(tmpdat)[1] <- "Gene"
write_tsv(tmpdat, path=paste0(filepath,"dengue_ribo_tpm.tsv"))

Dengue_RNA_TPM_25 <- round(Dengue_RNA_TPM_25, 4)
tmpdat <- as.data.frame(cbind(rownames(Dengue_RNA_TPM_25), Dengue_RNA_TPM_25))
colnames(tmpdat)[1] <- "Gene"
write_tsv(tmpdat, path=paste0(filepath,"dengue_rna_tpm.tsv"))

```