---
title: "Raw position-specific footprinting reads"
output: html_notebook
---

### Intialization and installation
```{r, echo=F, eval=F, message=F}
### Initialize packages
library(rhdf5)
library(GenomicRanges)
library(rtracklayer)
library(parallel)
library(Biostrings)
library(Rsamtools)
library(feather)
library(data.table)
library(GenomicAlignments)
library(tidyverse)
```

```{r}
### 3nt periodicity

datasets <- c("RiboNSC1","RiboNSC2","RiboNSC3","RibosiP1","RibosiP2","RibosiP3")
filepath <- "/data/riboseq/Dengue/output/shah/"


bcountData <- read_delim(paste0(filepath, "all_counts_bam.tsv"), 
    "\t", escape_double = FALSE, trim_ws = TRUE)
  
gff <- readGFFAsGRanges("/data/riboseq/Supp/seqs/Virus/Dengue/human_dengue.gff3")

genes <- as.character(bcountData$tid)
names(genes) <- as.character(bcountData$gene)
dataset <- "data"
out_data <- c()
```

```{r}
# Get gene and position specific total counts of all lengths
for(fn in datasets){
  ribo.filename <- paste0(filepath,fn,".h5")
  
  gene_sp_pos_counts <- mclapply(1:length(genes),
                                 function(x){
                                   gene <- genes[x]
                                   tmpdat <- h5read(file = ribo.filename, name = paste0("/",gene,"/",dataset,"/reads/data"))
                                   gid <- names(genes[x])
                                   tmpgff <- width(gff[seqnames(gff)==gene])
                                   tmpTib <- tibble(Gene = gid,
                                                    Transcript = gene,
                                                    UTR5 = tmpgff[1],
                                                    CDS = tmpgff[2],
                                                    UTR3 = tmpgff[3],
                                                    Pos = 1:ncol(tmpdat),
                                                    Counts = colSums(tmpdat[20:26,]))
                                   # tmpcounts <- c(gid, gene, tmpgff, colSums(tmpdat[20:26,]))
                        }, mc.cores = 40)

  a <- bind_rows(gene_sp_pos_counts)
  write_feather(a, paste0(filepath, fn, ".feather"))
}
```

### For frame and periodicity plots
```{r}
tmpData <- read_delim(paste0(filepath, datasets[1], ".tsv"), " ", na = c(" ","","NA"), col_names = F, guess_max = 1, col_types = list(col_character(), 
                                                                                                                                      col_character(),
                                                                                                                                      rep(col_integer(), 101521)))

tmpData <- read.table(paste0(filepath, datasets[1], ".tsv"), header = F, col.names = c("Gene", "Transcript", "utr5", "cds", "utr3", 1:101518), fill = T)
```








